### DISCUSSION TITLE
Can i use a separate channel (mask) to remove background and make that area transparent

### DISCUSSION THREAD

**Discussion by hanssonrickard** [libvips/libvips#2903] [Help wanted]
Hi.
I have a TIF file that is in CMYK colorspace and has a fifth channel that is a mask for the background to make it transparent.
The background itself is black.

Is it possible to use libvips to use the mask and remove the black area that is masked and make it transparent and save the TIFF so that in the end we have a clean 4 channel CMYK TIFF with transparent background instead of a mask?

Attached is also a screenshot of part of the file as well as the channels. The masked area is shown as red when the mask is active.
I had to crop the image as it is a production image from a customer that i cannot share in full size or show the full content.

I will use Python (pyvips) in this case.

vipsheader info for the testfile:

[root@ebe3e5763dbd data]# /usr/local/vips/bin/vipsheader -a ./TESTFILE_TRANSPARENS_AFTER_CS.tif 

(vipsheader:280): VIPS-WARNING **: 10:35:33.508: Sum of Photometric type-related color channels and ExtraSamples doesn't match SamplesPerPixel. Defining non-color channels as ExtraSamples.

(vipsheader:280): VIPS-WARNING **: 10:35:33.516: Sum of Photometric type-related color channels and ExtraSamples doesn't match SamplesPerPixel. Defining non-color channels as ExtraSamples.

(vipsheader:280): VIPS-WARNING **: 10:35:33.519: Sum of Photometric type-related color channels and ExtraSamples doesn't match SamplesPerPixel. Defining non-color channels as ExtraSamples.

(vipsheader:280): VIPS-WARNING **: 10:35:33.522: Sum of Photometric type-related color channels and ExtraSamples doesn't match SamplesPerPixel. Defining non-color channels as ExtraSamples.

(vipsheader:280): VIPS-WARNING **: 10:35:33.524: Sum of Photometric type-related color channels and ExtraSamples doesn't match SamplesPerPixel. Defining non-color channels as ExtraSamples.
./TESTFILE_TRANSPARENS_AFTER_CS.tif: 5504x8256 uchar, 5 bands, cmyk, tiffload
width: 5504
height: 8256
bands: 5
format: uchar
coding: none
interpretation: cmyk
xoffset: 0
yoffset: 0
xres: 11.811
yres: 11.811
filename: ./MODS_TRANSPARENS_AFTER_CS.tif
vips-loader: tiffload
n-pages: 1
icc-profile-data: 2194896 bytes of binary data
xmp-data: 20985 bytes of binary data
photoshop-data: 5812 bytes of binary data
resolution-unit: in
orientation: 1

<img width="1001" alt="screenshot" src="https://user-images.githubusercontent.com/4253602/176417098-db6076ec-2ee2-4196-939a-546faccb9fc0.png">

**Comment by jcupitt** [comment]
Hi @hanssonrickard,

Sure, just use `flatten` to press the alpha into the CMYK channels (if I understand your meaning). Could you upload a sample image?

**Reply by hanssonrickard** [reply]
Hi, no i wish to generate/keep transparency, but needs to knock out the black background using the mask that is in the channel named "Genomskinlighet" so that we get/see visible transparent background instead of the black background.

I can of course use the flatten to get a white background instead in this case, but can i also use flatten to keep/generate alpha?

**Reply by hanssonrickard** [reply]
Here is a part of the tif saved as new tif in photoshop.
Zipped it.

[PART_OF_TIFF.tif.zip](https://github.com/libvips/libvips/files/9010391/PART_OF_TIFF.tif.zip)

**Reply by jcupitt** [reply]
The terminology is very confusing :(

You want a CMYKA image, as before (so also five bands), but you want the K to be 0 in areas marked as transparent by A. Is that correct?

In Python you could do something like (untested):

```python
cmyka = pyvips.Image.new_from_file("PART_OF_TIFF.tif")
cmy = cmyka[0:3]
k = cmyka[3]
a = cmyka[4]
new_k = ((k * a) / 256).cast("uchar")
new_cmyka = cmy.bandjoin([new_k, a])
new_cmyka.write_to_file("x.tif")
```

**Reply by hanssonrickard** [reply]
Yes, that is correct. Sorry for confusing erminology when i tried to say what i wanted to do.
I will give it a try and see what the result is later this evening.

**Reply by hanssonrickard** [reply]
Hi.
Your example worked, it creates a new tif with the background transparent instead of black.
